### PR TITLE
Dmp

### PR DESCRIPTION
**Pull request by tchamelot**
In #18, I proposed to add DMP support for the MPU9250.

The pull request proposes to add a new mode `Dmp`. This mode allows use of the internal DMP, but it is not finished yet.

Currently, the `Dmp` mode supports:

- Programming the DMP either from an external firmware or from the firmware provided with the `dmp_firmware` feature (firmware taken from [librobotcontrol](https://github.com/StrawsonDesign/librobotcontrol))
- Enabling the Quaternion + Accelerometer + Gyroscope + Interruption features (hard coded)
- Reading the Quaternion from the internal FIFO

 and will support (planned):

- [x] Selecting the DMP features to enable (remove the hard coded ones)
- [x] Selecting the base orientation of the chip
- [ ] ~Supporting more features (tap detection)~ (will not be done)
- [x] Reading measurement more ergonomically
  - [x] Add a `all` method which depends on enabled features
  - [ ] ~Add `accel` and `gyro`  methods for `Dmp` mode~ (will not be done)

I have also provided an example for the BeagleBone Blue using the I2C bus. I did not test the feature with SPI as I am not used to it. If anyone could try to use the DMP with SPI and give me feedback it would be nice!

I would be happy to receive any feedback, suggestion or whatever!

Edit: Added check box for myself in order to keep up on the features that I have added but not yet pushed

### CODE REVIEW

**Comment by little-arhat** [comment]
Hey, @tchamelot, thanks for PR! 
I will have a look at the changeset, and will try to test it in the coming week.

**Comment by little-arhat** [comment]
Tried to build `semi_sensors` test from [proving-ground](https://github.com/copterust/proving-ground) with this branch:

```
error[E0599]: no method named `powi` found for type `f64` in the current scope
    --> /Users/r/prj/p/copterust/mpu9250/src/lib.rs:1109:47
     |
1109 |             quat.iter().map(|x| f64::from(*x).powi(2)).sum::<f64>().sqrt();
     |                                               ^^^^ method not found in `f64`

error[E0599]: no method named `sqrt` found for type `f64` in the current scope
    --> /Users/r/prj/p/copterust/mpu9250/src/lib.rs:1109:69
     |
1109 |             quat.iter().map(|x| f64::from(*x).powi(2)).sum::<f64>().sqrt();
     |                                                                     ^^^^ method not found in `f64`
```

For math operations we probably want to use [libm](https://github.com/rust-lang/libm)

**Comment by tchamelot** [comment]
I understand that `sqrt` and `pow` are not implemented in `no_std`. If you have any solution for the square root, I can use it. For the moment, I will only fix this by outputting non normalized quaternion.

**Comment by little-arhat** [comment]
@tchamelot I think we should put everything needed for dmp behind the feature-cfg, so for users, not interested in DMP nothing will change.
[sqrt](https://github.com/rust-lang/libm/blob/master/src/math/sqrt.rs) and [pow](https://github.com/rust-lang/libm/blob/master/src/math/pow.rs) are available in libm (we can use it as additional dependency, when dmp is enabled).

**Comment by tchamelot** [comment]
@little-arhat, I have passed most of dmp code behind the dmp feature. I  did not change anything i already wrote for configuration as it would be way more difficult to handle and it is invisible for the end user. I also removed the quaternion normalization and left it up to the end user.

File: src/conf.rs [hunk 1]
```diff
@@ -72,8 +72,9 @@ impl Default for AccelDataRate {
 impl AccelDataRate {
     pub(crate) fn accel_config_bits(&self) -> u8 {
         match self {
-            AccelDataRate::FChoice0 => 0b00010000,
-            AccelDataRate::DlpfConf(dlpf) => 0b00000000 | (*dlpf as u8),
+            AccelDataRate::FChoice0 => 0b0000_1000,
+            // 0x40 for 1024 bit fifo
```
Comment (little-arhat) [inline src/conf.rs:76]: Is it possible to split out formatting changes? Also, looks like rustfmt.toml committed to repo doesn't produce this change.

File: src/device.rs [hunk 2]
```diff
@@ -135,6 +143,21 @@ impl<SPI, NCS, E, EO> Device for SpiDevice<SPI, NCS>
         self.ncs.set_high().map_err(|a| SpiError::NCSError(a))?;
         Ok(())
     }
+
+    fn write_many(&mut self,
+                  reg: Register,
+                  buffer: &[u8])
+                  -> Result<(), Self::Error> {
+        if buffer.len() > 16 {
```
Comment (little-arhat) [inline src/device.rs:151]: Would be nice to handle this in compile-time, but I guess we have to wait for const generics or re-introduce genericarray crate to do that :-(

File: src/device.rs [hunk 3]
```diff
@@ -135,6 +143,21 @@ impl<SPI, NCS, E, EO> Device for SpiDevice<SPI, NCS>
...
+            return Err(Self::Error::WriteManyError);
```
Comment (little-arhat) [inline src/device.rs:152]: Not sure about this: `write_many` is not public, so error here always means library author error, failed assertion. But panicking here is even worse, so I guess we should sticki with this, until compile-time solution is ready.

File: src/device.rs [hunk 4]
```diff
@@ -94,6 +100,8 @@ pub enum SpiError<E, E2> {
     BusError(E),
     /// NCS error
     NCSError(E2),
+    /// Write many Error
```
Comment (little-arhat) [inline src/device.rs:103]: Maybe expand descrpition to show it's a library error?

File: src/lib.rs [hunk 5]
```diff
@@ -325,7 +378,10 @@ mod i2c_defs {
     {
         /// Creates a new [`Imu`] driver from an I2C peripheral
         /// with default configuration.
-        pub fn imu_default<D>(i2c: I2C, delay: &mut D) -> Result<Self, Error<E>>
+        pub fn imu_default<D>(
```
Comment (little-arhat) [inline src/lib.rs:381]: Maybe rustfmt change should be committed separately.

File: src/conf.rs (hunk 1 above)
Comment (tchamelot) [reply src/conf.rs:76]: Ok for the formatting. For the line 75, I have shifted the value to 1 to the left according to the documentation of the register map page 15 which says that `FChoice0` is bit 3. If you also want me to keep this change for latter tell me.

File: src/device.rs (hunk 2 above)
Comment (tchamelot) [reply src/device.rs:151]: So, are you ok to let this as is?

**Comment by tchamelot** [comment]
Hey, sorry for the long time without any update. I was working on something else. I have some time now to finish what I started. There will not be any more major change in this PR unless requested.

**Comment by tchamelot** [comment]
Hey @little-arhat , I am not happy with the `DmpMeasurement` composed with `Option`. Do you think that it would be a good idea to enable features at compile time. I could keep the `dmp` feature and add other features like `dmp_quat6` `dmp_quat` `dmp_accel` `dmp_gyro` `dmp_motion`. Those features would select the fields to keep in `DmpMeasurement` and simplify the parsing. However, the user would lose the possibility to use several mpu with different feature configurations.

File: src/device.rs (hunk 2 above)
Comment (little-arhat) [reply src/device.rs:151]: Yes!

**Comment by little-arhat** [comment]
Hey, thanks for the work, sorry for delay! 
Yeah, indeed, using features here might be to heavy handed. Let's use Options for now.

**Pull request merged**